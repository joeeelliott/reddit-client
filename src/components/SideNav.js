import React, { useState, useEffect } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useSelector, useDispatch } from 'react-redux'; 

import { selectInitialPostsState, searchPosts, userSearch, userNoSearch, showPosts, searchedPostsFound, selectIsSearching, selectSearch, selectToggleSideNav } from '../features/postSlice';

import PostFilters from './PostFilters'; 
import SpecificsFilters from './SpecificsFilters'; 

const SideNav = () => {
  const dispatch = useDispatch();
  const [searchText, setSearchText] = useState(""); 

  const postState = useSelector(selectInitialPostsState);
  const isSearching = useSelector(selectIsSearching);
  const search = useSelector(selectSearch);
  const toggleSideNav = useSelector(selectToggleSideNav); 
  
  const handleEyeMouseOver = (e) => {
    document.getElementsByClassName('sideNav_eye-icon-hover-text')[0].classList.add('sideNav_eye-icon-hover-text-show');    // executes animation 

    document.getElementsByClassName('sideNav_eye-icon-hover-text')[0].style.opacity = '1.0';   // keeps text visible whilst mouseover

    if(!postState.allPostsShown) {   // if there are hidden posts...
      document.getElementsByClassName('sideNav_eye-icon-hover-text')[0].innerHTML = 'Show hidden posts';   // ... change text
    }
  }

  const handleEyeMouseOut = (e) => {
    document.getElementsByClassName('sideNav_eye-icon-hover-text')[0].classList.remove('sideNav_eye-icon-hover-text-show');  // removes text

    document.getElementsByClassName('sideNav_eye-icon-hover-text')[0].style.opacity = '0';     // 0 opacity

    document.getElementsByClassName('sideNav_eye-icon-hover-text')[0].innerHTML = 'Show hidden posts';    // resets original text
  }

  const handleEyeClick = (e) => {
    if(postState.allPostsShown) {  // if no hidden posts
      document.getElementsByClassName('sideNav_eye-icon-hover-text')[0].innerHTML = 'No posts hidden';
    } else {  // if there are hidden posts
      dispatch(showPosts());    // show all hidden posts
      document.getElementsByClassName('sideNav_eye-icon-hover-text')[0].innerHTML = 'Posts unhidden';
    }
  }
  
  useEffect(() => {
    // console.log(searchText); 
    dispatch(searchPosts({ searchText }));   // executed in here as if executed in the handleChange, the userSearch string state is one letter behind what's in the search bar, so the search isn't accurate to what is in user search bar. 
  }, [searchText, dispatch]);

  useEffect(() => {
    !isSearching && setSearchText(""); 
    if(!searchText){
      dispatch(userNoSearch());
      dispatch(searchedPostsFound({ ids: undefined, text: searchText }));   // reset searchedPosts state array 
    }
  }, [dispatch, searchText, isSearching]);
  
  useEffect(() => {
    if(isSearching){
      dispatch(searchedPostsFound());
    }
  }, [dispatch, isSearching, search]);   // each time the searchText state changes in the store - not local state as that's always one change behind

  const handleChange = (e) => {
    setSearchText(e.currentTarget.value);
    if(e.currentTarget.value){
      dispatch(userSearch());
    } else {
      dispatch(userNoSearch());
    }
  }

  return (
    <div data-testid="sideNav-test" id="sideNav" className="sideNav_sideNav"> 
      <form className={!toggleSideNav ? "sideNav_hide-form" : undefined}>
        <label className="sideNav_primary-label" htmlFor="search">Search: </label>
        <input id="search" name="search" className="sideNav_search-input" type="text" value={searchText} placeholder="Search post title here..." onChange={handleChange} />
          
        <label className="sideNav_primary-label">Filter By: </label>

        <label className="sideNav_primary-label"><strong>Posts</strong></label>

        <PostFilters />

        <label className="sideNav_primary-label"><strong>Post Specifics</strong></label>

        <SpecificsFilters />

      </form>

      <div className="sideNav_eye-text-container">
        <div className="sideNav_text-container">
          <p className="sideNav_eye-icon-hover-text">Show hidden posts</p>
        </div>
        <div className="sideNav_eye-icon-container">
          <button 
            type="button" 
            aria-label="Restore any hidden posts back to the timeline" 
            onClick={handleEyeClick}
          >
            <FontAwesomeIcon 
              className="sideNav_eye-icon" 
              icon={['far', 'eye']} 
              onMouseOver={handleEyeMouseOver} 
              onMouseOut={handleEyeMouseOut} 
            />
          </button>
        </div>            
      </div>
    </div>
  )
}

export default SideNav;