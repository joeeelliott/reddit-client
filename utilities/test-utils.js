// THIS FILE IS FOR REFACTORED TEST FUNCTIONS THAT WILL BE KEPT IN A GLOBAL STATE AS THEY ARE LIKELY TO BE USED ACROSS MULTIPLE FILES

// See https://redux.js.org/usage/writing-tests for more info on below code. May be useful at a later date when using integration testing API calls with fetch, msw etc. This is setup, other code on there for example of the integration test. 

import React from 'react'; 
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom'; 
// Import your own reducer
import sideNavReducer from '../src/features/sideNavSlice';
import postReducer from '../src/features/postSlice'; 

function render(
  ui,
  {
    preloadedState,
    store = configureStore({ reducer: { sideNav: sideNavReducer, posts: postReducer }, preloadedState }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return ( 
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    )
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react'
//override render method
export { render }
