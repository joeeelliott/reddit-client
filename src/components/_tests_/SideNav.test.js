import React from 'react';
// import { Provider } from 'react-redux';
// import { render as rtlRender, screen, fireEvent, cleanup } from '@testing-library/react'; 
// import { store } from '../../app/store';

import { render, cleanup, screen } from '../../../utilities/test-utils';

import SideNav from '../SideNav'; 

// const render = component => rtlRender(
//   <Provider store={store}>   
//     {component}
//   </Provider>
// );

describe('SideNav Component', () => {

  // let component;
  // beforeEach(() => {
  //   component = render(<SideNav />);
  // });

  afterEach(cleanup); 

  it('renders without crashing', () => {   
    render(<SideNav />);
    screen.debug(undefined, 300000)
  });

  // it('renders as expected; matches snapshot', () => {       
  //   expect(screen).toMatchSnapshot();
  //   expect(component).toMatchSnapshot();
  // });

  // can unit test my input fields/confirm button etc. here
});