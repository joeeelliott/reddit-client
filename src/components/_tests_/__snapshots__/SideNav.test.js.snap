// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav Component renders as expected; matches snapshot 1`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`SideNav Component renders as expected; matches snapshot 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sideNav_hide-nav"
        data-testid="sideNav"
        role="sideNav-outer-div"
      >
        <form
          role="form"
        >
          <label
            role="label"
          >
            Search: 
          </label>
          <input
            placeholder="Search term here..."
            role="input"
            type="text"
          />
          <label
            role="label"
          >
            Filter: 
          </label>
          <select
            id="filters"
            name="filters"
            role="select"
          >
            <option
              value=""
            />
            <option
              value=""
            />
            <option
              value=""
            />
            <option
              value=""
            />
          </select>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-filter fa-w-16 "
            data-icon="filter"
            data-prefix="fas"
            data-test-header_filter-icon="true"
            data-test-header_font-awesome-icon="true"
            data-testid="filter-icon"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
              fill="currentColor"
            />
          </svg>
          <div
            class="sideNav_btn-container"
            role="btn-container"
          >
            <button
              class="sideNav_btn"
            >
              Confirm
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sideNav_hide-nav"
      data-testid="sideNav"
      role="sideNav-outer-div"
    >
      <form
        role="form"
      >
        <label
          role="label"
        >
          Search: 
        </label>
        <input
          placeholder="Search term here..."
          role="input"
          type="text"
        />
        <label
          role="label"
        >
          Filter: 
        </label>
        <select
          id="filters"
          name="filters"
          role="select"
        >
          <option
            value=""
          />
          <option
            value=""
          />
          <option
            value=""
          />
          <option
            value=""
          />
        </select>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-filter fa-w-16 "
          data-icon="filter"
          data-prefix="fas"
          data-test-header_filter-icon="true"
          data-test-header_font-awesome-icon="true"
          data-testid="filter-icon"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
            fill="currentColor"
          />
        </svg>
        <div
          class="sideNav_btn-container"
          role="btn-container"
        >
          <button
            class="sideNav_btn"
          >
            Confirm
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
